From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "yinwei.sun1" <yinwei.sun1@unisoc.com>
Date: Thu, 22 Dec 2022 15:42:26 +0800
Subject: [PATCH 118/226] Use getPackagesForOps instead of iterating packages
 by checkOperation

In some low end device, avoid iterating packages by checkOperation can
optimize ~100ms.

Bug:258771872
Test: manual
Change-Id: I13dc0b8c4a6a07fcf3663ddf0fe8d8c6599c9756
---
 .../android/server/StorageManagerService.java | 29 ++++++++++++-------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/services/core/java/com/android/server/StorageManagerService.java b/services/core/java/com/android/server/StorageManagerService.java
index 81e70659f0..78c2434081 100644
--- a/services/core/java/com/android/server/StorageManagerService.java
+++ b/services/core/java/com/android/server/StorageManagerService.java
@@ -186,6 +186,7 @@ import java.io.PrintWriter;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
@@ -2193,19 +2194,27 @@ class StorageManagerService extends IStorageManager.Stub
     private void snapshotAndMonitorLegacyStorageAppOp(UserHandle user) {
         int userId = user.getIdentifier();
 
-        // TODO(b/149391976): Use mIAppOpsService.getPackagesForOps instead of iterating below
-        // It should improve performance but the AppOps method doesn't return any app here :(
-        // This operation currently takes about ~20ms on a freshly flashed device
+        List<AppOpsManager.PackageOps> pkgs = null;
+        try {
+            pkgs = mIAppOpsService.getPackagesForOps(new int[] { OP_LEGACY_STORAGE });
+        } catch(RemoteException e) {
+            Slog.e(TAG, "Failed to getPackagesForOps", e);
+        }
+        Set<String> legacyStoragePackages = new HashSet<>();
+        if (pkgs != null) {
+            for (AppOpsManager.PackageOps pkg : pkgs) {
+                for (AppOpsManager.OpEntry op : pkg.getOps()) {
+                    if (op.getMode() == MODE_ALLOWED) {
+                        legacyStoragePackages.add(pkg.getPackageName());
+                    }
+                }
+            }
+        }
         for (ApplicationInfo ai : mPmInternal.getInstalledApplications(MATCH_DIRECT_BOOT_AWARE
                         | MATCH_DIRECT_BOOT_UNAWARE | MATCH_UNINSTALLED_PACKAGES | MATCH_ANY_USER,
                         userId, Process.myUid())) {
-            try {
-                boolean hasLegacy = mIAppOpsService.checkOperation(OP_LEGACY_STORAGE, ai.uid,
-                        ai.packageName) == MODE_ALLOWED;
-                updateLegacyStorageApps(ai.packageName, ai.uid, hasLegacy);
-            } catch (RemoteException e) {
-                Slog.e(TAG, "Failed to check legacy op for package " + ai.packageName, e);
-            }
+            boolean hasLegacy = legacyStoragePackages.contains(ai.packageName);
+            updateLegacyStorageApps(ai.packageName, ai.uid, hasLegacy);
         }
 
         if (mPackageMonitorsForUser.get(userId) == null) {
